Sources/Mantle/Substitution.swift

```diff
@@ -0,0 +1,286 @@
+/// Substitution.swift
+///
+/// Copyright 2017, The Silt Language Project.
+///
+/// This project is released under the MIT license, a copy of which is
+/// available in the repository.
+
+import Moho
+
+/// A substitution `σ : Δ → Γ` can be seen as a list of terms `Γ ⊢ vᵢ : Aᵢ`
+/// with `Aᵢ` from `Δ`, that can be applied to a term `Δ ⊢ uᵢ : Bᵢ` yielding
+/// `Γ ⊢ uᵢσ : Bᵢσ` by substituting each `vᵢ` for a variable in `uᵢ`.
+public struct Substitution {
```

Comment:
NB: This representation is going to be abhorrently slow.  Change it instead to be a wrapper around `[Internal]` and remove the recursive parameters.